floors/fetcher.go

```diff
@@ -0,0 +1,294 @@
+	}
+
+	// Check for floors JSON in cache
+	result, found := f.Get(config.Fetch.URL)
+	if found {
+		fetcherResult, ok := result.(*openrtb_ext.PriceFloorRules)
+		if !ok || fetcherResult.Data == nil {
+			return nil, openrtb_ext.FetchError
+		}
+		return fetcherResult, openrtb_ext.FetchSuccess
+	}
+
+	//miss: push to channel to fetch and return empty response
+	if config.Enabled && config.Fetch.Enabled && config.Fetch.Timeout > 0 {
+		fetchInfo := FetchInfo{AccountFloorFetch: config.Fetch, FetchTime: time.Now().Unix(), RefetchRequest: false}
+		f.configReceiver <- fetchInfo
+	}
+
+	return nil, openrtb_ext.FetchInprogress
+}
+
+type FetchQueue []*FetchInfo
+
+func (fq FetchQueue) Len() int {
+	return len(fq)
+}
+
+func (fq FetchQueue) Less(i, j int) bool {
+	return fq[i].FetchTime < fq[j].FetchTime
+}
+
+func (fq FetchQueue) Swap(i, j int) {
+	fq[i], fq[j] = fq[j], fq[i]
+}
+
+func (fq *FetchQueue) Push(element interface{}) {
+	fetchInfo := element.(*FetchInfo)
+	*fq = append(*fq, fetchInfo)
+}
+
+func (fq *FetchQueue) Pop() interface{} {
+	old := *fq
+	n := len(old)
+	fetchInfo := old[n-1]
+	old[n-1] = nil
+	*fq = old[0 : n-1]
+	return fetchInfo
+}
+
+func (fq *FetchQueue) Top() *FetchInfo {
+	old := *fq
+	if len(old) == 0 {
+		return nil
+	}
+	return old[0]
+}
+
+func workerPanicHandler(p interface{}) {
+	glog.Errorf("floor fetcher worker panicked: %v", p)
+}
+
+func NewPriceFloorFetcher(maxWorkers, maxCapacity, cacheCleanUpInt, cacheExpiry int) *PriceFloorFetcher {
+	floorFetcher := PriceFloorFetcher{
+		pool:            pond.New(maxWorkers, maxCapacity, pond.PanicHandler(workerPanicHandler)),
+		fetchQueue:      make(FetchQueue, 0, 100),
+		fetchInProgress: make(map[string]bool),
+		configReceiver:  make(chan FetchInfo, maxCapacity),
+		done:            make(chan struct{}),
+		cacheExpiry:     time.Duration(cacheExpiry) * time.Second,
+		cache:           cache.New(time.Duration(cacheExpiry)*time.Second, time.Duration(cacheCleanUpInt)*time.Second),
+	}
+
+	go floorFetcher.Fetcher()
+
+	return &floorFetcher
+}
+
+func (f *PriceFloorFetcher) SetWithExpiry(key string, value interface{}, cacheExpiry time.Duration) {
+	f.cache.Set(key, value, cacheExpiry)
+}
+
+func (f *PriceFloorFetcher) Get(key string) (interface{}, bool) {
+	return f.cache.Get(key)
+}
+
+func (f *PriceFloorFetcher) worker(configs config.AccountFloorFetch) {
+	floorData, fetchedMaxAge := fetchAndValidate(configs)
+	if floorData != nil {
+		// Update cache with new floor rules
+		glog.Infof("Updating Value in cache for URL %s", configs.URL)
+		cacheExpiry := f.cacheExpiry
+		if fetchedMaxAge != 0 && fetchedMaxAge > configs.Period && fetchedMaxAge < math.MaxInt32 {
+			cacheExpiry = time.Duration(fetchedMaxAge) * time.Second
+		}
+		f.SetWithExpiry(configs.URL, floorData, cacheExpiry)
+	}
+
+	// Send to refetch channel
+	f.configReceiver <- FetchInfo{AccountFloorFetch: configs, FetchTime: time.Now().Add(time.Duration(configs.Period) * time.Second).Unix(), RefetchRequest: true}
```

Comment:
For testing, please extract all calls to `time.Now()` with our existing `timeutil.Time` interface.

Reply:
Done